### ISSUE TITLE
Application restarts when the `io.balena.features.gpu` label is set

### ISSUE THREAD

**Issue by alexgg** [balena-io/balena-supervisor#1449]
I am using a jetson-tx2 on 2.47.0+rev2 with a v11.12.4 supervisor.

Using the test application in https://github.com/balena-io-playground/idle-multiapp, I observe the following:

* If I remove the `io.balena.features.gpu` label from the compose file both idle apps run fine
* If I add the `io.balena.features.gpu` to an application, it will enter a restart loop with a ~9s cadence

The behaviour occues both if the label is in only one app or in both apps. The apps with the label will keep restarting.

**Comment by robertgzr** [comment]
@alexgg does the `2.47.0+rev2` image include the nvidie-container-toolkit stuff? if not this is expected...

**Comment by alexgg** [comment]
@robertgzr I could expect the GPU label not to work withouth the nvidia-container-tookit stuff, but why does the container app enter in a restart loop?

**Comment by robertgzr** [comment]
@alexgg because docker will attempt to find a GPU device plugin which it doesn't find and then fail to start the container... and then the supervisor will retry to spawn the container and we have this situation. I can imagine something like a warning that the container toolkit stuff isn't available but that's more a UX thing, isn't it?

**Comment by pgils** [comment]
@robertgzr I can replicate this exact problem *with toolkit installed* as well.

Device is `photon-xavier-nx` v2.58.3+rev6 (from [this branch](https://github.com/hwky-ai/balena-jetson/tree/nvidia-gpu-test)) with supervisor v11.14.0. 

Full additional package list: [packagegroup-balena-nvidia-gpu](https://github.com/hwky-ai/balena-jetson/commit/84c40742a5c310ab35d0e0130b86278d41a5dcb1)

Containers run fine (with GPU access and libraries mounted) when started with `balena-cli` on the device e.g:
```
# balena run -ti --gpus 1 --net host hwkyai/photon-xavier-nx-deepstream-demo:32.4.3
```
Using `balena push` with the `io.balena.features.gpu = 1` label set results in the same ~9s restart loop @alexgg observed.

**Comment by 20k-ultra** [comment]
Has anyone checked what the exit code is for the container ? Is it crashing or receiving a kill signal ? @pgils @alexgg 

If it's receiving a kill signal I'd suspect the Supervisor but if it's crashing then it seems like an OS issue.

**Comment by 20k-ultra** [comment]
We have seen that on a device with this label the Supervisor is logging an issue when trying to start:

```
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [event]   Event: Service start {"service":{"appId":1783841,"serviceId":876396,"serviceName":"examples","releaseId":1660536}}
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]   Scheduling another update attempt in 64000ms due to failure:  Error: Failed to apply state transition steps. (HTTP code 500) server error - could not select device driver "" with capabilities: [[gpu]]  Steps:["start"]
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]         at fn (/usr/src/app/dist/app.js:6:8488)
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]       at runMicrotasks (<anonymous>)
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]       at processTicksAndRejections (internal/process/task_queues.js:97:5)
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]   Device state apply error Error: Failed to apply state transition steps. (HTTP code 500) server error - could not select device driver "" with capabilities: [[gpu]]  Steps:["start"]
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]         at fn (/usr/src/app/dist/app.js:6:8488)
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]       at runMicrotasks (<anonymous>)
Jan 15 16:25:54 6f4bfdd resin-supervisor[4166]: [error]       at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

@robertgzr and I have made some notes with questions we'd like to follow up with and will have more info soon.

**Comment by pgils** [comment]
Is this on a recent version of `resin-supervisor` @20k-ultra ?
On `2.56.0+rev1` for `photon-nano` with supervisor `v11.14` this is the error i get:
```
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]   Scheduling another update attempt in 32000ms due to failure:  Error: Failed to apply state transition steps. (HTTP code 500) server error - OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: initialization error: driver error: failed to process request\\\\n\\\"\"": unknown  Steps:["start"]
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]         at /usr/src/app/dist/app.js:10:8684
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]       at runMicrotasks (<anonymous>)
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]       at processTicksAndRejections (internal/process/task_queues.js:97:5)
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]   Device state apply error Error: Failed to apply state transition steps. (HTTP code 500) server error - OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: initialization error: driver error: failed to process request\\\\n\\\"\"": unknown  Steps:["start"]
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]         at /usr/src/app/dist/app.js:10:8684
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]       at runMicrotasks (<anonymous>)
Jan 18 11:14:03 68aaad1 resin-supervisor[4118]: [error]       at processTicksAndRejections (internal/process/task_queues.js:97:5)
```
[balena-stock-full.log](https://github.com/balena-io/balena-supervisor/files/5830189/balena-stock-full.log)

Locally built `2.60.1+rev2` + `tegra-libraries`/`toolkit` with Supervisor `v11.14` ([branch](https://github.com/hwky-ai/balena-jetson/tree/nvidia-gpu-test)):
```
Jan 18 11:53:11 balena resin-supervisor[3201]: [event]   Event: Service start {"service":{"appId":1,"serviceId":1,"serviceName":"ubuntu-gpu","releaseId":1}}
Jan 18 11:53:13 balena resin-supervisor[3201]: [event]   Event: Service started {"service":{"appId":1,"serviceId":1,"serviceName":"ubuntu-gpu","releaseId":1}}
Jan 18 11:53:13 balena resin-supervisor[3201]: [api]     GET /v2/state/status 200 - 1670.175 ms
Jan 18 11:53:14 balena resin-supervisor[3201]: [api]     GET /v2/state/status 200 - 139.485 ms
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]   Replacing container for service ubuntu-gpu because of config changes:
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]     Non-array fields:  {"added":{},"deleted":{"deviceRequests":{"0":{}}},"updated":{}}
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]   Replacing container for service ubuntu-gpu because of config changes:
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]     Non-array fields:  {"added":{},"deleted":{"deviceRequests":{"0":{}}},"updated":{}}
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]   Replacing container for service ubuntu-gpu because of config changes:
Jan 18 11:53:14 balena resin-supervisor[3201]: [debug]     Non-array fields:  {"added":{},"deleted":{"deviceRequests":{"0":{}}},"updated":{}}
Jan 18 11:53:14 balena resin-supervisor[3201]: [event]   Event: Service kill {"service":{"appId":1,"serviceId":1,"serviceName":"ubuntu-gpu","releaseId":1}}
Jan 18 11:53:15 balena resin-supervisor[3201]: [api]     GET /v2/state/status 200 - 150.343 ms
Jan 18 11:53:15 balena resin-supervisor[3201]: [debug]   Replacing container for service ubuntu-gpu because of config changes:
Jan 18 11:53:15 balena resin-supervisor[3201]: [debug]     Non-array fields:  {"added":{},"deleted":{"deviceRequests":{"0":{}}},"updated":{}}
Jan 18 11:53:15 balena resin-supervisor[3201]: [debug]   Replacing container for service ubuntu-gpu because of config changes:
Jan 18 11:53:15 balena resin-supervisor[3201]: [debug]     Non-array fields:  {"added":{},"deleted":{"deviceRequests":{"0":{}}},"updated":{}}
```
[balena-toolkit-gpu.log](https://github.com/balena-io/balena-supervisor/files/5830196/balena-toolkit-gpu.log)

**Comment by pgils** [comment]
I found the contents of the `deviceRequests` array to be inequal here:

https://github.com/balena-io/balena-supervisor/blob/e6ebfca0b93ad5992ba87d4d2edaed0257ac48c2/src/compose/service.ts#L756-L759

`thisOmitted`:
```
deviceRequests: [
    {
      Driver: '',
      Count: 1,
      DeviceIDs: null,
      Capabilities: [Array],
      Options: null
    }
  ],
```
`otherOmitted`:
```
deviceRequests: [ { Count: 1, Capabilities: [Array] } ],
```
Are uninitialized values stripped somewhere?

Ommitting `deviceRequests` from the comparison makes the label work as expected for me: https://github.com/pgils/balena-supervisor/commit/c779243b4acdcdcc71930cdc0882261156fb14a6

This now runs on GPU:
```
version: '2'
services:
  gpu-test:
    image: hwkypgils/photon-nano-jetson-tftrt-demo:32.4.3
    privileged: true
    labels:
      - "io.balena.features.gpu=1"
```

**Comment by 20k-ultra** [comment]
Awesome @pgils! Thank you for digging this up.

@robertgzr or myself will check it out soon to confirm so we can get this resolved.

**Comment by alexgg** [comment]
This is fixed in supervisor v12.3.6. See https://github.com/balena-io/balena-supervisor/pull/1582.
